File: app/static/js/scores.js
// Score Entry JavaScript (Dual Mode) v3
console.log("Scores JS Loaded");

let entryMode = 'student'; // 'student' or 'subject'
let selectedFilters = {
    session: null,
    term: null,
    class: null,
    student: null,
    subject: null
};

// Helper to safe-get value
const getValue = (id) => {
    const el = document.getElementById(id);
    return el ? el.value : null;
};

document.addEventListener('DOMContentLoaded', async function () {
    const urlParams = new URLSearchParams(window.location.search);
    const urlStudentId = urlParams.get('student_id');
    const urlClassId = urlParams.get('class_id');

    if (urlClassId) {
        document.getElementById('class-select').value = urlClassId;
        selectedFilters.class = urlClassId;
        await loadClassStudents(urlClassId);
        await loadClassSubjects(urlClassId);

        if (urlStudentId) {
            document.getElementById('student-select').value = urlStudentId;
            selectedFilters.student = urlStudentId;
            loadGrid();
        }
    }
});

// Mode Toggling
document.querySelectorAll('.btn-mode').forEach(btn => {
    btn.onclick = function () {
        document.querySelectorAll('.btn-mode').forEach(b => b.classList.remove('active'));
        this.classList.add('active');
        entryMode = this.dataset.mode;

        // UI Updates
        document.getElementById('student-filter-group').style.display = entryMode === 'student' ? 'block' : 'none';
        document.getElementById('subject-filter-group').style.display = entryMode === 'subject' ? 'block' : 'none';
        document.getElementById('row-header').textContent = entryMode === 'student' ? 'Subject' : 'Student';

        clearGrid();
        if (entryMode === 'subject' && selectedFilters.class) {
            loadClassSubjects(selectedFilters.class);
        }
    };
});

// Session Change -> Load Terms
document.getElementById('session-select').addEventListener('change', async function () {
    const sessionId = this.value;
    selectedFilters.session = sessionId;
    const termSelect = document.getElementById('term-select');

    termSelect.innerHTML = '<option value="">Loading...</option>';

    if (!sessionId) {
        termSelect.innerHTML = '<option value="">Select Term</option>';
        return;
    }

    try {
        const response = await fetch(`/scores/api/terms/${sessionId}`);
        const terms = await response.json();

        termSelect.innerHTML = '<option value="">Select Term</option>';
        terms.forEach(term => {
            const option = document.createElement('option');
            option.value = term.id;
            option.textContent = `Term ${term.term_number}`;
            termSelect.appendChild(option);
        });
    } catch (e) {
        console.error("Error loading terms:", e);
    }
    clearGrid();
});

// Class Change -> Load Students AND Subjects
document.getElementById('class-select').addEventListener('change', async function () {
    const classId = this.value;
    selectedFilters.class = classId;

    loadClassStudents(classId);
    loadClassSubjects(classId);
    clearGrid();
});

async function loadClassStudents(classId) {
    const studentSelect = document.getElementById('student-select');
    if (!classId) {
        studentSelect.innerHTML = '<option value="">Select Student</option>';
        return;
    }

    studentSelect.innerHTML = '<option value="">Loading students...</option>';
    try {
        const response = await fetch(`/scores/api/students_in_class/${classId}`);
        const students = await response.json();

        studentSelect.innerHTML = '<option value="">Select Student</option>';
        students.forEach(std => {
            const option = document.createElement('option');
            option.value = std.id;
            option.textContent = `${std.last_name} ${std.first_name} (${std.reg_number})`;
            studentSelect.appendChild(option);
        });
    } catch (e) {
        studentSelect.innerHTML = '<option value="">Error loading students</option>';
    }
}

async function loadClassSubjects(classId) {
    const subjectSelect = document.getElementById('subject-select');
    if (!classId) {
        subjectSelect.innerHTML = '<option value="">Select Subject</option>';
        return;
    }

    subjectSelect.innerHTML = '<option value="">Loading subjects...</option>';
    try {
        const response = await fetch(`/scores/api/class_subjects/${classId}`);
        const subjects = await response.json();

        subjectSelect.innerHTML = '<option value="">Select Subject</option>';
        subjects.forEach(sub => {
            const option = document.createElement('option');
            option.value = sub.id;
            option.textContent = sub.name;
            subjectSelect.appendChild(option);
        });
    } catch (e) {
        subjectSelect.innerHTML = '<option value="">Error loading subjects</option>';
    }
}

// Student/Subject Selection -> Load Grid
document.getElementById('student-select').onchange = () => {
    selectedFilters.student = getValue('student-select');
    loadGrid();
};

document.getElementById('subject-select').onchange = () => {
    selectedFilters.subject = getValue('subject-select');
    loadGrid();
};

document.getElementById('term-select').onchange = () => {
    selectedFilters.term = getValue('term-select');
    loadGrid();
};

async function loadGrid() {
    if (entryMode === 'student') {
        loadStudentScoreGrid();
    } else {
        loadSubjectScoreGrid();
    }
}

async function loadStudentScoreGrid() {
    const tbody = document.getElementById('score-tbody');
    const sessionId = getValue('session-select');
    const termId = getValue('term-select');
    const classId = getValue('class-select');
    const studentId = getValue('student-select');

    if (!sessionId || !termId || !classId || !studentId) return;

    tbody.innerHTML = '<tr><td colspan="9" class="no-data">Loading scores...</td></tr>';

    const params = new URLSearchParams({ session_id: sessionId, term_id: termId, class_id: classId, student_id: studentId });

    try {
        const response = await fetch(`/scores/api/student_score_grid?${params}`);
        const subjects = await response.json();

        if (subjects.length === 0) {
            tbody.innerHTML = '<tr><td colspan="9" class="no-data">No subjects assigned to this class</td></tr>';
            return;
        }

        tbody.innerHTML = subjects.map(sub => `
            <tr data-subject-id="${sub.subject_id}">
                <td style="font-weight: 500;">${sub.subject_name}</td>
                <td><input type="number" class="ca-input form-control" value="${sub.ca}" min="0" max="30"></td>
                <td><input type="number" class="exam-input form-control" value="${sub.exam}" min="0" max="70"></td>
                <td class="readonly total-cell" style="font-weight: bold;">${sub.total}</td>
                <td class="readonly grade-cell" style="color: var(--primary); font-weight: bold;">${sub.grade}</td>
                <td class="readonly text-muted">${sub.position}</td>
                <td class="readonly text-muted">${sub.highest}</td>
                <td class="readonly text-muted">${sub.lowest}</td>
                <td class="readonly text-muted">${sub.average}</td>
            </tr>
        `).join('');

        addInputListeners();
    } catch (e) {
        tbody.innerHTML = `<tr><td colspan="9" class="no-data">Error: ${e.message}</td></tr>`;
    }
}

async function loadSubjectScoreGrid() {
    const tbody = document.getElementById('score-tbody');
    const sessionId = getValue('session-select');
    const termId = getValue('term-select');
    const classId = getValue('class-select');
    const subjectId = getValue('subject-select');

    if (!sessionId || !termId || !classId || !subjectId) return;

    tbody.innerHTML = '<tr><td colspan="9" class="no-data">Loading students...</td></tr>';

    const params = new URLSearchParams({ session_id: sessionId, term_id: termId, class_id: classId, subject_id: subjectId });

    try {
        const response = await fetch(`/scores/api/subject_score_grid?${params}`);
        const students = await response.json();

        if (students.length === 0) {
            tbody.innerHTML = '<tr><td colspan="9" class="no-data">No students found in this class</td></tr>';
            return;
        }

        tbody.innerHTML = students.map(std => `
            <tr data-student-id="${std.student_id}">
                <td style="font-weight: 500;"><div>${std.full_name}</div><small class="text-muted">${std.reg_number}</small></td>
                <td><input type="number" class="ca-input form-control" value="${std.ca}" min="0" max="30"></td>
                <td><input type="number" class="exam-input form-control" value="${std.exam}" min="0" max="70"></td>
                <td class="readonly total-cell" style="font-weight: bold;">${std.total}</td>
                <td class="readonly grade-cell" style="color: var(--primary); font-weight: bold;">${std.grade}</td>
                <td class="readonly text-muted">${std.position}</td>
                <td class="readonly text-muted">${std.highest}</td>
                <td class="readonly text-muted">${std.lowest}</td>
                <td class="readonly text-muted">${std.average}</td>
            </tr>
        `).join('');

        addInputListeners();
    } catch (e) {
        tbody.innerHTML = `<tr><td colspan="9" class="no-data">Error: ${e.message}</td></tr>`;
    }
}

function clearGrid() {
    document.getElementById('score-tbody').innerHTML = '<tr><td colspan="9" class="no-data">Select filters to load grid</td></tr>';
}

function addInputListeners() {
    document.querySelectorAll('.ca-input, .exam-input').forEach(input => {
        input.oninput = function () {
            const row = this.closest('tr');
            let ca = parseFloat(row.querySelector('.ca-input').value) || 0;
            let exam = parseFloat(row.querySelector('.exam-input').value) || 0;
            if (ca > 30) { ca = 30; row.querySelector('.ca-input').value = 30; }
            if (exam > 70) { exam = 70; row.querySelector('.exam-input').value = 70; }
            const total = ca + exam;
            row.querySelector('.total-cell').textContent = total;
            row.querySelector('.grade-cell').textContent = calculateGrade(total);
        };
    });
}

function calculateGrade(total) {
    if (total >= 80) return 'A1';
    if (total >= 70) return 'B2';
    if (total >= 65) return 'B3';
    if (total >= 60) return 'C4';
    if (total >= 55) return 'C5';
    if (total >= 50) return 'C6';
    if (total >= 45) return 'D7';
    if (total >= 40) return 'E8';
    return 'F9';
}

// SAVE
async function saveScores() {
    const btn = document.getElementById('save-btn');
    const headerBtn = document.getElementById('header-save-btn');
    const originalText = btn.innerHTML;

    btn.disabled = headerBtn.disabled = true;
    btn.innerHTML = headerBtn.innerHTML = '<i class="fas fa-spinner fa-spin"></i> Saving...';

    const session_id = getValue('session-select');
    const term_id = getValue('term-select');
    const class_id = getValue('class-select');

    const rows = document.querySelectorAll('#score-tbody tr');
    const scores = [];

    rows.forEach(row => {
        const id = entryMode === 'student' ? row.dataset.subjectId : row.dataset.studentId;
        if (id) {
            const ca = parseFloat(row.querySelector('.ca-input').value) || 0;
            const exam = parseFloat(row.querySelector('.exam-input').value) || 0;
            const entry = { ca, exam };
            if (entryMode === 'student') entry.subject_id = id;
            else entry.student_id = id;
            scores.push(entry);
        }
    });

    const url = entryMode === 'student' ? '/scores/api/save_student_scores' : '/scores/api/save_subject_scores';
    const body = { session_id, term_id, class_id, scores };
    if (entryMode === 'student') body.student_id = getValue('student-select');
    else body.subject_id = getValue('subject-select');

    try {
        const response = await fetch(url, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(body)
        });
        const result = await response.json();
        alert(result.message);
        loadGrid();
    } catch (e) {
        alert('Error: ' + e.message);
    } finally {
        btn.disabled = headerBtn.disabled = false;
        btn.innerHTML = originalText;
        headerBtn.innerHTML = '<i class="fas fa-save"></i> Save Scores';
    }
}

document.getElementById('save-btn').onclick = saveScores;
document.getElementById('header-save-btn').onclick = saveScores;

// --- Excel Export / Import ---
document.getElementById('export-btn').onclick = () => {
    const classId = getValue('class-select');
    const sessionId = getValue('session-select');
    const termId = getValue('term-select');

    if (!classId || !sessionId || !termId) {
        alert('Please select Session, Term, and Class first.');
        return;
    }

    let url = `/scores/api/export_excel?class_id=${classId}&session_id=${sessionId}&term_id=${termId}`;
    if (entryMode === 'subject') {
        const subjectId = getValue('subject-select');
        if (subjectId) url += `&subject_id=${subjectId}`;
    }

    window.location.href = url;
};

document.getElementById('import-btn').onclick = () => {
    const classId = getValue('class-select');
    if (!classId) {
        alert('Please select a class first.');
        return;
    }
    document.getElementById('import-file').click();
};

document.getElementById('import-file').onchange = async function () {
    if (!this.files[0]) return;

    const file = this.files[0];
    const formData = new FormData();
    formData.append('file', file);
    formData.append('class_id', getValue('class-select'));
    formData.append('session_id', getValue('session-select'));
    formData.append('term_id', getValue('term-select'));

    const btn = document.getElementById('import-btn');
    const originalText = btn.innerHTML;
    btn.innerHTML = '<i class="fas fa-spinner fa-spin"></i> Importing...';
    btn.disabled = true;

    try {
        const res = await fetch('/scores/api/import_excel', {
            method: 'POST',
            body: formData
        });
        const result = await res.json();

        if (result.success) {
            alert(result.message);
            loadGrid();
        } else {
            alert('Import Failed: ' + result.message);
            if (result.errors) console.error(result.errors);
        }
    } catch (e) {
        alert('Error: ' + e.message);
    } finally {
        btn.innerHTML = originalText;
        btn.disabled = false;
        this.value = ''; // Reset input
    }
};

File: app/routes/scores.py
"""Score Entry Routes"""
from flask import Blueprint, render_template, request, jsonify
import sys
import os
sys.path.insert(0, os.path.join(os.path.dirname(os.path.dirname(os.path.dirname(__file__))), 'src'))
from database.db_manager import DatabaseManager
from logic.grading import calculate_grade

bp = Blueprint('scores', __name__, url_prefix='/scores')
db = DatabaseManager()

@bp.route('/')
def index():
    """Score entry page"""
    sessions = db.execute_query('SELECT * FROM sessions ORDER BY name DESC')
    classes = db.execute_query('SELECT * FROM classes ORDER BY level, stream')
    # Subjects are loaded dynamically based on class selection via JS
    return render_template('scores/index.html', sessions=sessions, classes=classes)

@bp.route('/api/terms/<int:session_id>')
def get_terms(session_id):
    """Get terms for a session"""
    terms = db.execute_query('SELECT * FROM terms WHERE session_id = ? ORDER BY term_number', (session_id,))
    return jsonify([dict(t) for t in terms])

@bp.route('/api/students_in_class/<int:class_id>')
def get_students_in_class(class_id):
    """Get all active students in a class"""
    print(f"Fetching students for class_id: {class_id}")
    # Single line query to ensure matching
    students = db.execute_query(
        'SELECT id, first_name, last_name, reg_number FROM students WHERE class_id = ? AND active_status = 1 ORDER BY last_name, first_name', 
        (class_id,)
    )
    print(f"Found {len(students)} students")
    return jsonify([dict(s) for s in students])

@bp.route('/api/student_score_grid')
def get_student_score_grid():
    """Get score grid for a single student (Rows=Subjects)"""
    student_id = request.args.get('student_id', type=int)
    class_id = request.args.get('class_id', type=int)
    session_id = request.args.get('session_id', type=int)
    term_id = request.args.get('term_id', type=int)
    
    if not all([student_id, class_id, session_id, term_id]):
        return jsonify([])
        
    # 1. Get all subjects assigned to this class
    class_subjects = db.execute_query('''
        SELECT s.id, s.name, s.code 
        FROM subjects s
        JOIN class_subjects cs ON s.id = cs.subject_id
        WHERE cs.class_id = ?
        ORDER BY cs.display_order, s.name
    ''', (class_id,))
    
    result = []
    for sub in class_subjects:
        # 2. Get existing score for this student/subject
        score = db.execute_query('''
            SELECT * FROM scores 
            WHERE student_id = ? AND subject_id = ? AND session_id = ? AND term_id = ?
        ''', (student_id, sub['id'], session_id, term_id))
        
        score_data = score[0] if score else {}
        
        result.append({
            'subject_id': sub['id'],
            'subject_name': sub['name'],
            'ca': score_data.get('ca_score', 0),
            'exam': score_data.get('exam_score', 0),
            'total': score_data.get('total', 0),
            'grade': score_data.get('grade', '-'),
            'position': score_data.get('position', '-'),
            'highest': score_data.get('class_highest', '-'),
            'lowest': score_data.get('class_lowest', '-'),
            'average': score_data.get('class_average', '-')
        })
        
    return jsonify(result)

@bp.route('/api/subject_score_grid')
def get_subject_score_grid():
    """Get score grid for all students in a class for one subject (Rows=Students)"""
    class_id = request.args.get('class_id', type=int)
    subject_id = request.args.get('subject_id', type=int)
    session_id = request.args.get('session_id', type=int)
    term_id = request.args.get('term_id', type=int)
    
    if not all([class_id, subject_id, session_id, term_id]):
        return jsonify([])
        
    # 1. Get all students in this class
    students = db.execute_query('''
        SELECT id, first_name, last_name, reg_number 
        FROM students 
        WHERE class_id = ? AND active_status = 1
        ORDER BY last_name, first_name
    ''', (class_id,))
    
    result = []
    for std in students:
        # 2. Get existing score for this student/subject
        score = db.execute_query('''
            SELECT * FROM scores 
            WHERE student_id = ? AND subject_id = ? AND session_id = ? AND term_id = ?
        ''', (std['id'], subject_id, session_id, term_id))
        
        score_data = score[0] if score else {}
        
        result.append({
            'student_id': std['id'],
            'full_name': f"{std['last_name']} {std['first_name']}",
            'reg_number': std['reg_number'],
            'ca': score_data.get('ca_score', 0),
            'exam': score_data.get('exam_score', 0),
            'total': score_data.get('total', 0),
            'grade': score_data.get('grade', '-'),
            'position': score_data.get('position', '-'),
            'highest': score_data.get('class_highest', '-'),
            'lowest': score_data.get('class_lowest', '-'),
            'average': score_data.get('class_average', '-')
        })
        
    return jsonify(result)

@bp.route('/api/save_subject_scores', methods=['POST'])
def save_subject_scores():
    """Save scores for all students in a class for one subject"""
    data = request.json
    class_id = data.get('class_id')
    subject_id = data.get('subject_id')
    session_id = data.get('session_id')
    term_id = data.get('term_id')
    scores_data = data.get('scores', [])

    if not all([class_id, subject_id, session_id, term_id]):
        return jsonify({'success': False, 'message': 'Missing required fields'}), 400

    try:
        saved_count = 0
        for entry in scores_data:
            student_id = entry['student_id']
            ca = float(entry['ca'])
            exam = float(entry['exam'])
            total = ca + exam
            grade = calculate_grade(total)

            db.execute_update('''
                INSERT OR REPLACE INTO scores 
                (student_id, subject_id, session_id, term_id, ca_score, exam_score, total, grade)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?)
            ''', (student_id, subject_id, session_id, term_id, ca, exam, total, grade))
            saved_count += 1

        # Trigger Ranking
        ranking_message = ""
        try:
            from logic.ranking import RankingEngine
            engine = RankingEngine()
            res = engine.process_class_results(class_id, session_id, term_id)
            if not res['success']:
                print(f"Ranking warning: {res['message']}")
                ranking_message = f" (Ranking warning: {res['message']})"
        except Exception as re:
            print(f"Ranking failed: {re}")
            ranking_message = " (Ranking failed to update)"

        return jsonify({'success': True, 'message': f'Scores saved for {saved_count} students{ranking_message}'})
    except Exception as e:
        return jsonify({'success': False, 'message': str(e)}), 500

@bp.route('/api/class_subjects/<int:class_id>')
def get_class_subjects(class_id):
    """Get subjects assigned to a class"""
    subjects = db.execute_query('''
        SELECT s.id, s.name, s.code 
        FROM subjects s
        JOIN class_subjects cs ON s.id = cs.subject_id
        WHERE cs.class_id = ?
        ORDER BY s.name
    ''', (class_id,))
    return jsonify([dict(s) for s in subjects])

@bp.route('/api/save_student_scores', methods=['POST'])
def save_student_scores():
    """Save scores for a single student (multiple subjects) with validation"""
    data = request.json
    student_id = data.get('student_id')
    class_id = data.get('class_id')
    session_id = data.get('session_id')
    term_id = data.get('term_id')
    scores_data = data.get('scores', [])

    if not all([student_id, class_id, session_id, term_id]):
        return jsonify({
            'success': False,
            'message': 'Missing required fields'
        }), 400

    if not scores_data:
        return jsonify({
            'success': False,
            'message': 'No scores provided'
        }), 400

    try:
        # Validate all scores first
        for score_entry in scores_data:
            ca_score = float(score_entry['ca'])
            exam_score = float(score_entry['exam'])

            # Range validation
            if not (0 <= ca_score <= 30):
                return jsonify({
                    'success': False,
                    'message': f'CA score must be between 0 and 30 (got {ca_score})'
                }), 400

            if not (0 <= exam_score <= 70):
                return jsonify({
                    'success': False,
                    'message': f'Exam score must be between 0 and 70 (got {exam_score})'
                }), 400

        # Save all scores
        saved_count = 0
        for score_entry in scores_data:
            subject_id = score_entry['subject_id']
            ca_score = float(score_entry['ca'])
            exam_score = float(score_entry['exam'])
            total = ca_score + exam_score
            grade = calculate_grade(total)

            db.execute_update('''
                INSERT OR REPLACE INTO scores
                (student_id, subject_id, session_id, term_id, ca_score, exam_score, total, grade)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?)
            ''', (student_id, subject_id, session_id, term_id, ca_score, exam_score, total, grade))
            saved_count += 1

        # Trigger Ranking Calculation
        ranking_success = False
        ranking_msg = ""
        try:
            from logic.ranking import RankingEngine
            engine = RankingEngine()
            res = engine.process_class_results(class_id, session_id, term_id)
            if res['success']:
                ranking_success = True
            else:
                ranking_msg = res['message']
                print(f"Ranking warning: {ranking_msg}")
        except Exception as re_err:
            print(f"Ranking update failed: {re_err}")
            ranking_msg = str(re_err)

        # Fetch updated scores with rankings
        updated_scores = []
        if ranking_success:
            for score_entry in scores_data:
                subject_id = score_entry['subject_id']
                score = db.execute_query('''
                    SELECT * FROM scores
                    WHERE student_id = ? AND subject_id = ? AND session_id = ? AND term_id = ?
                ''', (student_id, subject_id, session_id, term_id))

                if score:
                    updated_scores.append({
                        'subject_id': subject_id,
                        'position': score[0].get('position'),
                        'class_highest': score[0].get('class_highest'),
                        'class_lowest': score[0].get('class_lowest'),
                        'class_average': score[0].get('class_average')
                    })
        
        msg = f'{saved_count} scores saved'
        if not ranking_success:
             msg += f' (Ranking issue: {ranking_msg})'

        return jsonify({
            'success': True,
            'message': msg,
            'updated_scores': updated_scores,
            'ranking_success': ranking_success
        })

    except ValueError as ve:
        return jsonify({
            'success': False,
            'message': f'Invalid score value: {str(ve)}'
        }), 400
    except Exception as e:
        return jsonify({
            'success': False,
            'message': f'Failed to save scores: {str(e)}'
        }), 500

@bp.route('/api/save', methods=['POST'])
def save_scores():
    """Save scores (Legacy/Subject-Centric)"""
    data = request.json
    class_id = data.get('class_id')
    subject_id = data.get('subject_id')
    session_id = data.get('session_id')
    term_id = data.get('term_id')
    scores_data = data.get('scores', [])
    
    if not all([class_id, subject_id, session_id, term_id]):
        return jsonify({'success': False, 'message': 'Missing required fields'})
    
    try:
        for score_entry in scores_data:
            student_id = score_entry['student_id']
            ca_score = float(score_entry['ca'])
            exam_score = float(score_entry['exam'])
            total = ca_score + exam_score
            grade = calculate_grade(total)
            
            db.execute_update('''
                INSERT OR REPLACE INTO scores 
                (student_id, subject_id, session_id, term_id, ca_score, exam_score, total, grade)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?)
            ''', (student_id, subject_id, session_id, term_id, ca_score, exam_score, total, grade))
        
        return jsonify({'success': True, 'message': 'Scores saved successfully!'})
    except Exception as e:
        return jsonify({'success': False, 'message': str(e)})

# --- Excel Export / Import ---

@bp.route('/api/export_excel')
def export_excel():
    """Export scores to Excel"""
    import openpyxl
    from openpyxl.styles import Font, Alignment
    from io import BytesIO
    from flask import send_file

    try:
        class_id = request.args.get('class_id')
        session_id = request.args.get('session_id')
        term_id = request.args.get('term_id')
        subject_id = request.args.get('subject_id') # Optional (for subject mode)

        if not all([class_id, session_id, term_id]):
            return "Missing parameters", 400

        # Fetch Context
        class_info = db.execute_query('SELECT name FROM classes WHERE id=?', (class_id,))[0]
        session_info = db.execute_query('SELECT name FROM sessions WHERE id=?', (session_id,))[0]
        term_info = db.execute_query('SELECT term_number FROM terms WHERE id=?', (term_id,))[0]
        
        output = BytesIO()
        workbook = openpyxl.Workbook()
        sheet = workbook.active
        sheet.title = "Scores"

        # Headers
        headers = ['Reg Number', 'Last Name', 'First Name', 'Subject Code', 'Subject Name', 'CA (30)', 'Exam (70)', 'Total']
        sheet.append(headers)
        
        # Style Headers
        for cell in sheet[1]:
            cell.font = Font(bold=True)
            cell.alignment = Alignment(horizontal='center')

        # Fetch Data
        if subject_id:
            # Export for specific subject (All students in class)
            subject = db.execute_query('SELECT code, name FROM subjects WHERE id=?', (subject_id,))[0]
            students = db.execute_query('SELECT id, reg_number, last_name, first_name FROM students WHERE class_id=? AND active_status=1 ORDER BY last_name', (class_id,))
            
            for std in students:
                score = db.execute_query('SELECT ca_score, exam_score FROM scores WHERE student_id=? AND subject_id=? AND session_id=? AND term_id=?',
                                        (std['id'], subject_id, session_id, term_id))
                ca = score[0]['ca_score'] if score else 0
                exam = score[0]['exam_score'] if score else 0
                
                sheet.append([
                    std['reg_number'], std['last_name'], std['first_name'], 
                    subject['code'], subject['name'], 
                    ca, exam, ca + exam
                ])
        else:
            # Export all subjects for all students (Big Grid)
            # For simplicity, we'll list rows normally: Student - Subject - Score
            students = db.execute_query('SELECT id, reg_number, last_name, first_name FROM students WHERE class_id=? AND active_status=1 ORDER BY last_name', (class_id,))
            subjects = db.execute_query('SELECT s.id, s.code, s.name FROM subjects s JOIN class_subjects cs ON s.id=cs.subject_id WHERE cs.class_id=? ORDER BY s.name', (class_id,))
            
            for std in students:
                for sub in subjects:
                    score = db.execute_query('SELECT ca_score, exam_score FROM scores WHERE student_id=? AND subject_id=? AND session_id=? AND term_id=?',
                                            (std['id'], sub['id'], session_id, term_id))
                    ca = score[0]['ca_score'] if score else 0
                    exam = score[0]['exam_score'] if score else 0
                    
                    sheet.append([
                        std['reg_number'], std['last_name'], std['first_name'], 
                        sub['code'], sub['name'], 
                        ca, exam, ca + exam
                    ])

        # Auto-adjust column width
        for col in sheet.columns:
            max_length = 0
            column = col[0].column_letter # Get the column name
            for cell in col:
                try:
                    if len(str(cell.value)) > max_length:
                        max_length = len(str(cell.value))
                except:
                    pass
            adjusted_width = (max_length + 2)
            sheet.column_dimensions[column].width = adjusted_width

        workbook.save(output)
        output.seek(0)
        
        filename = f"Scores_{class_info['name']}_{session_info['name']}_Term{term_info['term_number']}.xlsx".replace('/', '-')
        return send_file(output, as_attachment=True, download_name=filename, mimetype='application/vnd.openxmlformats-officedocument.spreadsheetml.sheet')

    except Exception as e:
        print(f"Export Error: {e}")
        return str(e), 500

@bp.route('/api/import_excel', methods=['POST'])
def import_excel():
    """Import scores from Excel"""
    import openpyxl
    
    if 'file' not in request.files:
        return jsonify({'success': False, 'message': 'No file uploaded'})
    
    file = request.files['file']
    class_id = request.form.get('class_id')
    session_id = request.form.get('session_id')
    term_id = request.form.get('term_id')
    
    if not all([file, class_id, session_id, term_id]):
         return jsonify({'success': False, 'message': 'Missing parameters'})

    try:
        wb = openpyxl.load_workbook(file)
        sheet = wb.active
        
        # Expected Headers: Reg Number | Last Name | First Name | Subject Code | Subject Name | CA | Exam
        # We allow flexible column indices
        headers = {}
        for cell in sheet[1]:
            headers[cell.value.lower().strip()] = cell.col_idx - 1 # 0-based index
            
        required = ['reg number', 'subject code', 'ca (30)', 'exam (70)']
        for req in required:
            if req not in headers:
                 # Try minimal matching
                 if req == 'ca (30)' and 'ca' in headers: headers['ca (30)'] = headers['ca']
                 elif req == 'exam (70)' and 'exam' in headers: headers['exam (70)'] = headers['exam']
                 else:
                    return jsonify({'success': False, 'message': f'Missing column: {req}'})

        success_count = 0
        errors = []
        
        for row_idx, row in enumerate(sheet.iter_rows(min_row=2, values_only=True), start=2):
            try:
                reg_num = row[headers['reg number']]
                sub_code = row[headers['subject code']]
                ca = row[headers['ca (30)']] or 0
                exam = row[headers['exam (70)']] or 0
                
                if not reg_num or not sub_code: continue
                
                # Validation
                try:
                    ca = float(ca)
                    exam = float(exam)
                except:
                    errors.append(f"Row {row_idx}: Invalid number format")
                    continue
                    
                if ca > 30 or exam > 70:
                    errors.append(f"Row {row_idx}: Scores out of range (CA<=30, Exam<=70)")
                    continue

                # Resolve IDs
                std = db.execute_query('SELECT id FROM students WHERE reg_number = ?', (reg_num,))
                if not std:
                    errors.append(f"Row {row_idx}: Student {reg_num} not found")
                    continue
                
                sub = db.execute_query('SELECT id FROM subjects WHERE code = ?', (sub_code,))
                if not sub:
                    errors.append(f"Row {row_idx}: Subject {sub_code} not found")
                    continue
                    
                # Save
                total = ca + exam
                grade = calculate_grade(total)
                
                db.execute_update('''
                    INSERT OR REPLACE INTO scores 
                    (student_id, subject_id, session_id, term_id, ca_score, exam_score, total, grade)
                    VALUES (?, ?, ?, ?, ?, ?, ?, ?)
                ''', (std[0]['id'], sub[0]['id'], session_id, term_id, ca, exam, total, grade))
                
                success_count += 1
                
            except Exception as row_err:
                errors.append(f"Row {row_idx}: {str(row_err)}")

        # Rankings update
        ranking_msg = ""
        try:
             res = ranking_engine.process_class_results(class_id, session_id, term_id)
        except:
             pass

        return jsonify({
            'success': True,
            'message': f'Imported {success_count} scores. {len(errors)} errors.',
            'errors': errors[:10] # Top 10 errors
        })
        
    except Exception as e:
        return jsonify({'success': False, 'message': f"File error: {str(e)}"})
